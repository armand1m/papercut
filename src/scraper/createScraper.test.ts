import { orderBy } from 'lodash';
import path from 'path';

process.env.PAPERCUT_PAGE_CACHE_PATH = path.resolve(
  __dirname,
  './__fixtures__/testpagecache'
);

process.env.PAPERCUT_GEOSEARCH_CACHE_PATH = path.resolve(
  __dirname,
  './__fixtures__/testgeocache'
);

test('createScraper - Single page (strict mode off)', async () => {
  const { createScraper } = await import('./createScraper');
  const scraper = createScraper({
    name: `Hacker News`,
    options: {
      log: false,
      cache: true,
    },
  });

  const results = await scraper.run({
    strict: false,
    baseUrl: 'https://news.ycombinator.com/',
    target: '.athing',
    selectors: {
      rank: (utils) => {
        const value = utils.text('.rank').replace(/^\D+/g, '');
        return Number(value);
      },
      name: ({ text }) => text('.titlelink'),
      url: ({ href }) => href('.titlelink'),
      score: ({ element }) => {
        return element.nextElementSibling?.querySelector('.score')
          ?.textContent;
      },
      createdBy: ({ element }) => {
        return element.nextElementSibling?.querySelector('.hnuser')
          ?.textContent;
      },
      createdAt: ({ element }) => {
        return element.nextElementSibling
          ?.querySelector('.age')
          ?.getAttribute('title');
      },
      failingOnPurpose: () => {
        throw new Error('nope');
      },
    },
  });

  expect(results).toMatchSnapshot();
});

test('createScraper - Single page (strict mode on)', async () => {
  const { createScraper } = await import('./createScraper');
  const scraper = createScraper({
    name: `Hacker News`,
    options: {
      log: false,
      cache: true,
    },
  });

  const results = await scraper.run({
    strict: true,
    baseUrl: 'https://news.ycombinator.com/',
    target: '.athing',
    selectors: {
      rank: (utils) => {
        const value = utils.text('.rank').replace(/^\D+/g, '');
        return Number(value);
      },
      name: ({ text }) => text('.titlelink'),
      url: ({ href }) => href('.titlelink'),
      score: ({ element }) => {
        return element.nextElementSibling?.querySelector('.score')
          ?.textContent;
      },
      createdBy: ({ element }) => {
        return element.nextElementSibling?.querySelector('.hnuser')
          ?.textContent;
      },
      createdAt: ({ element }) => {
        return element.nextElementSibling
          ?.querySelector('.age')
          ?.getAttribute('title');
      },
    },
  });

  expect(results).toMatchSnapshot();
});

test('createScraper - Pagination', async () => {
  jest.setTimeout(120_000);
  const { createScraper } = await import('./createScraper');

  const createLabeledUrl = (label: string, url: string) => ({
    label,
    url,
  });

  const scraper = createScraper({
    name: 'Amsterdam Coffeeshops',
    options: {
      cache: true,
    },
  });

  const results = await scraper.run({
    strict: true,
    target: '.summary-box',
    baseUrl:
      'https://amsterdamcoffeeshops.com/search/item/coffeeshops',
    pagination: {
      enabled: true,
      lastPageNumberSelector:
        '.navigation > .pagination > li:nth-child(8) > a',
      createPaginatedUrl: (baseUrl, pageNumber) => {
        return `${baseUrl}/p:${pageNumber}`;
      },
    },
    selectors: {
      name: ({ text }) => {
        return text('.media-body > h3 > a');
      },
      description: ({ text }) => {
        return text('.media-body > .summary-desc');
      },
      photo: ({ src }) => {
        return { url: src('.media-left > a > img') };
      },
      phone: ({ text }) => {
        return text('.media-right > .contact-info > mark > a');
      },
      address: ({ text }) => {
        const address = text('.media-body > address > p');

        if (!address) {
          return undefined;
        }

        return address.replace(/\s+/g, ' ').replace(/^\s+|\s+$/g, '');
      },
      location: async (selectors, $this) => {
        const address = $this.address(selectors, $this);
        return selectors.geosearch(address);
      },
      social: ({ href }) => {
        const websiteHref = href('.visit-website');
        return websiteHref
          ? [createLabeledUrl('Official Website', websiteHref)]
          : [];
      },
      menus: () => {
        /** TODO: scrape menus */
        return [];
      },
      badges: ({ all }) => {
        const { asArray: badges } = all(
          '.media-left > div > div > img'
        );

        return badges
          .map((badge) => badge.getAttribute('title'))
          .filter((badge) => badge !== undefined) as string[];
      },
      rating: ({ className }) => {
        const rateNumber = className(
          '.media-right > .summary-info > span > span'
        );

        if (!rateNumber) {
          return 0;
        }

        return Number(rateNumber.replace('rate-', ''));
      },
    },
  });

  const sortedResults = orderBy(
    results,
    [(coffeeshop) => coffeeshop.name.toLowerCase()],
    ['asc']
  );

  expect(sortedResults).toMatchSnapshot();
});
