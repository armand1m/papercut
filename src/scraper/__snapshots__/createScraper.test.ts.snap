// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createScraper 1`] = `
Array [
  Object {
    "createdAt": "2021-11-14T00:15:27",
    "createdBy": "mancerayder",
    "name": "U.S. states file updated antitrust complaint against Google",
    "rank": 1,
    "score": "127 points",
    "url": "https://www.reuters.com/world/us/us-states-file-updated-antitrust-complaint-against-alphabets-google-2021-11-13/",
  },
  Object {
    "createdAt": "2021-11-13T22:47:42",
    "createdBy": "todsacerdoti",
    "name": "Hoax email blast abused poor coding in FBI website",
    "rank": 2,
    "score": "184 points",
    "url": "https://krebsonsecurity.com/2021/11/hoax-email-blast-abused-poor-coding-in-fbi-website/",
  },
  Object {
    "createdAt": "2021-11-13T20:50:57",
    "createdBy": "thunderbong",
    "name": "The case of the 500-mile email (2002)",
    "rank": 3,
    "score": "201 points",
    "url": "https://www.ibiblio.org/harris/500milemail.html",
  },
  Object {
    "createdAt": "2021-11-13T17:18:12",
    "createdBy": "picture",
    "name": "Reverse-engineering the Yamaha DX7 synthesizer's sound chip from die photos",
    "rank": 4,
    "score": "311 points",
    "url": "https://www.righto.com/2021/11/reverse-engineering-yamaha-dx7.html",
  },
  Object {
    "createdAt": "2021-11-13T22:41:06",
    "createdBy": "dootah",
    "name": "Study finds a difference between neurons of humans and other mammals",
    "rank": 5,
    "score": "64 points",
    "url": "https://news.mit.edu/2021/neurons-humans-mammals-1110",
  },
  Object {
    "createdAt": "2021-11-13T21:47:36",
    "createdBy": "bobitsaboy",
    "name": "Dataminers are finding developer comments and unlicensed songs in GTA Trilogy",
    "rank": 6,
    "score": "73 points",
    "url": "https://www.pcgamer.com/uk/dataminers-are-finding-developer-comments-and-unlicensed-songs-in-gta-trilogy-as-rockstar-games-launcher-remains-offline/",
  },
  Object {
    "createdAt": "2021-11-13T21:14:11",
    "createdBy": "jkuria",
    "name": "In a ‘learning trap’ experiment, adults leap conclusions while children explore",
    "rank": 7,
    "score": "87 points",
    "url": "https://www.wsj.com/articles/why-children-learn-better-than-adults-11636033185",
  },
  Object {
    "createdAt": "2021-11-13T23:19:58",
    "createdBy": "elsewhen",
    "name": "Richard ‘Lowtax’ Kyanka, founder of Something Awful, has died",
    "rank": 8,
    "score": "43 points",
    "url": "https://www.vice.com/en/article/v7dmqb/richard-lowtax-kyanka-founder-of-something-awful-is-dead-at-45",
  },
  Object {
    "createdAt": "2021-11-13T20:36:00",
    "createdBy": "rmason",
    "name": "How Michigan grew its startup ecosystem",
    "rank": 9,
    "score": "95 points",
    "url": "https://timesofe.com/how-michigan-grew-its-startup-ecosystem-and-became-the-state-with-the-fastest-growing-venture-capital/",
  },
  Object {
    "createdAt": "2021-11-13T13:04:37",
    "createdBy": "IceWreck",
    "name": "Restic – Backups Done Right",
    "rank": 10,
    "score": "424 points",
    "url": "https://restic.net/",
  },
  Object {
    "createdAt": "2021-11-13T22:31:59",
    "createdBy": "tomohawk",
    "name": "Vaccum-chamber quantum sensor device could allow GPS-free navigation",
    "rank": 11,
    "score": "53 points",
    "url": "https://breakingdefense.com/2021/11/sandias-atomic-avocado-could-allow-gps-free-pnt/",
  },
  Object {
    "createdAt": "2021-11-12T13:21:50",
    "createdBy": "johanznsdn",
    "name": "Show HN: Stringmark – Chrome extension that makes online research easier",
    "rank": 12,
    "score": "68 points",
    "url": "https://www.stringmark.com",
  },
  Object {
    "createdAt": "2021-11-12T04:55:42",
    "createdBy": "healeycodes",
    "name": "Spotify Codes – Part 2",
    "rank": 13,
    "score": "36 points",
    "url": "https://boonepeter.github.io/posts/spotify-codes-part-2/",
  },
  Object {
    "createdAt": "2021-11-13T21:26:19",
    "createdBy": "throwaway9838",
    "name": "Ask HN: Advice for moving on from a failed startup?",
    "rank": 14,
    "score": "59 points",
    "url": "item?id=29213305",
  },
  Object {
    "createdAt": "2021-11-13T15:56:36",
    "createdBy": "ali92hm",
    "name": "Breaking down the 'payment for order flow' debate",
    "rank": 15,
    "score": "17 points",
    "url": "https://a16z.com/2021/02/17/payment-for-order-flow/",
  },
  Object {
    "createdAt": "2021-11-12T20:31:47",
    "createdBy": "xenodium",
    "name": "Show HN: Plain Org – a new org mode app for iOS",
    "rank": 16,
    "score": "32 points",
    "url": "https://apps.apple.com/us/app/plain-org/id1578965002",
  },
  Object {
    "createdAt": "2021-11-13T17:03:47",
    "createdBy": "todsacerdoti",
    "name": "Managing an external display on Linux shouldn’t be this hard",
    "rank": 17,
    "score": "68 points",
    "url": "https://changelog.complete.org/archives/10311-managing-an-external-display-on-linux-shouldnt-be-this-hard",
  },
  Object {
    "createdAt": "2021-11-13T21:00:51",
    "createdBy": undefined,
    "name": "Substack (YC W18) is hiring engineering managers to build the future of writing",
    "rank": 18,
    "score": undefined,
    "url": "https://jobs.lever.co/substackinc/ae724ae2-7882-4ac6-ab70-699327f8f2e7",
  },
  Object {
    "createdAt": "2021-11-12T04:21:44",
    "createdBy": "sblank",
    "name": "When there seems to be no way out – customer discovery for your head",
    "rank": 19,
    "score": "42 points",
    "url": "https://steveblank.com/2021/11/11/when-there-seems-to-be-no-way-out-customer-discovery-for-your-head/",
  },
  Object {
    "createdAt": "2021-11-13T03:53:59",
    "createdBy": "tjalfi",
    "name": "The weirdest bug I've ever encountered",
    "rank": 20,
    "score": "81 points",
    "url": "https://mental-reverb.com/blog.php?id=29",
  },
  Object {
    "createdAt": "2021-11-13T14:49:12",
    "createdBy": "malshe",
    "name": "In the 17th century, Leibniz dreamed of a machine that could calculate ideas",
    "rank": 21,
    "score": "91 points",
    "url": "https://spectrum.ieee.org/in-the-17th-century-leibniz-dreamed-of-a-machine-that-could-calculate-ideas",
  },
  Object {
    "createdAt": "2021-11-13T14:52:08",
    "createdBy": "krycroft",
    "name": "Show HN: Marvin – A grumpy Slackbot who doesn't like your ideas",
    "rank": 22,
    "score": "94 points",
    "url": "https://github.com/keltrycroft/marvin",
  },
  Object {
    "createdAt": "2021-11-13T06:01:19",
    "createdBy": "Jtsummers",
    "name": "Metaobject protocols: Why we want them and what else they can do (1993) [pdf]",
    "rank": 23,
    "score": "38 points",
    "url": "https://users.cs.duke.edu/~vahdat/ps/mop.pdf",
  },
  Object {
    "createdAt": "2021-11-10T20:40:46",
    "createdBy": "joeyespo",
    "name": "The strong and weak forces of architecture",
    "rank": 24,
    "score": "25 points",
    "url": "https://martinfowler.com/articles/strong-weak-arch.html",
  },
  Object {
    "createdAt": "2021-11-13T17:33:39",
    "createdBy": "tosh",
    "name": "Project Euler",
    "rank": 25,
    "score": "251 points",
    "url": "https://projecteuler.net",
  },
  Object {
    "createdAt": "2021-11-13T18:05:15",
    "createdBy": "nixcraft",
    "name": "Dte: A language for expressing and calculating date and time",
    "rank": 26,
    "score": "48 points",
    "url": "https://github.com/mvrozanti/dte",
  },
  Object {
    "createdAt": "2021-11-12T13:15:08",
    "createdBy": "nz",
    "name": "Gin, television, and social surplus, or, “looking for the mouse” (2008)",
    "rank": 27,
    "score": "96 points",
    "url": "https://gist.github.com/jm3/6724931",
  },
  Object {
    "createdAt": "2021-11-13T11:37:06",
    "createdBy": "duiker101",
    "name": "Show HN: Relative Bookmarks Browser Extension",
    "rank": 28,
    "score": "22 points",
    "url": "https://github.com/duiker101/relative_bookmarks",
  },
  Object {
    "createdAt": "2021-11-13T09:13:49",
    "createdBy": "h1x",
    "name": "Sign arbitrary data with your SSH keys",
    "rank": 29,
    "score": "548 points",
    "url": "https://www.agwa.name/blog/post/ssh_signatures",
  },
  Object {
    "createdAt": "2021-11-13T13:14:18",
    "createdBy": "theduder99",
    "name": "Chemists discover new way to harness energy from ammonia",
    "rank": 30,
    "score": "55 points",
    "url": "https://phys.org/news/2021-11-chemists-harness-energy-ammonia.html",
  },
]
`;
